File: loader/src/main/java/com/blackmorse/hattrick/promotions/PromotionListsMerger.java
package com.blackmorse.hattrick.promotions;

import com.blackmorse.hattrick.promotions.model.IPromotion;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.function.BiFunction;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

/**
 * Generics for Unit Tests
 */
public class PromotionListsMerger<Team, Promote extends IPromotion> {
    private final BiFunction<List<Team>, List<Team>, Promote> mapFunction;

    public PromotionListsMerger(BiFunction<List<Team>, List<Team>, Promote> mapFunction) {
        this.mapFunction = mapFunction;
    }

    public List<Promote> mergeTeams(List<Team> leftList, List<Team> rightList) {
        List<Promote> result = new ArrayList<>();

        int i = 0;
        while (i < leftList.size()) {
            int leftMinInd = i;
            int leftMaxInd = i;
            while(leftMaxInd < leftList.size() && leftList.get(leftMinInd).equals(leftList.get(leftMaxInd))) {
                leftMaxInd++;
            }
            leftMaxInd--;

            int rightMinInd = i;
            while(rightMinInd >=0 && rightList.get(rightMinInd).equals(rightList.get(i))) {
                rightMinInd--;
            }
            rightMinInd++;

            int rightMaxInd = leftMaxInd;
            while(rightMaxInd < rightList.size() && rightList.get(rightMaxInd).equals(rightList.get(leftMaxInd))) {
                rightMaxInd++;
            }
            rightMaxInd--;

            List<Team> leftArray = IntStream.range(leftMinInd, leftMaxInd + 1)
                    .mapToObj(leftList::get)
                    .collect(Collectors.toList());

            List<Team> rightArray = IntStream.range(rightMinInd, rightMaxInd + 1)
                    .mapToObj(rightList::get)
                    .collect(Collectors.toList());


            result.add(mapFunction.apply(leftArray, rightArray));

            i = leftMaxInd + 1;
        }

        Iterator<Promote> iterator = result.iterator();

        Promote previous = iterator.next();
        Promote next ;//= //terator.next();
        do {
            next = iterator.next();
            if (previous.getUpTeams().equals(next.getUpTeams())) {
                iterator.remove();
                previous.getDownTeams().addAll(next.getDownTeams());
            } else {
                previous = next;
            }
//            next = iterator.next();
        } while (iterator.hasNext());

//        if (previous.getUpTeams().equals(next.getUpTeams())) {
//            iterator.remove();
//            previous.getDownTeams().addAll(next.getDownTeams());
//        }

            return result;
    }
}

File: akka-loader/src/main/scala/loadergraph/promotions/PromotionListsMerger.scala
package loadergraph.promotions

import models.clickhouse.PromotionModelCH
import models.stream.StreamTeam

import java.util
import scala.collection.mutable.ArrayBuffer
import scala.jdk.CollectionConverters.{IterableHasAsScala, IteratorHasAsScala, MutableSeqHasAsJava}

object PromotionListsMerger {
  def merge(leagueId: Int, season: Int, upDivisionLevel: Int, promoteType: PromoteType,
             leftList: List[StreamTeam], rightList: List[StreamTeam]): List[PromotionModelCH] = {
    val result = ArrayBuffer[PromotionModelCH]()

    var i = 0
    while (i < leftList.size) {
      val leftMinInd = i

      var leftMaxInd = i
      while (leftMaxInd < leftList.size && (leftList(leftMinInd) samePosition leftList(leftMaxInd))) leftMaxInd += 1
      leftMaxInd -= 1

      var rightMinInd = i
      while (rightMinInd >= 0 && (rightList(rightMinInd) samePosition rightList(i))) rightMinInd -= 1
      rightMinInd += 1

      var rightMaxInd = leftMaxInd
      while (rightMaxInd < rightList.size && (rightList(rightMaxInd) samePosition rightList(leftMaxInd))) rightMaxInd += 1
      rightMaxInd -= 1

      val leftArray = (leftMinInd to leftMaxInd).map(leftList(_))

      val rightArray = (rightMinInd to rightMaxInd).map(rightList(_))


      val promotionModel = PromotionModelCH(leagueId = leagueId,
        season = season,
        upDivisionLevel = upDivisionLevel,
        promoteType = promoteType,
        downTeams = leftArray.toBuffer,
        upTeams = rightArray.toList
      )

      result += promotionModel

      i = leftMaxInd + 1
    }

    val javaResult = new util.ArrayList(result.asJava)
    val iterator = javaResult.iterator

    var previous = iterator.next
    var next: PromotionModelCH = null

    do {
      next = iterator.next
      if (previous.upTeams == next.upTeams) {
        iterator.remove()
        previous.addDownTeams(next.downTeams)
      }
      else previous = next
    } while (iterator.hasNext)

    javaResult.asScala.toList
  }
}
